### PR TITLE
Refactors Github Workflows

### PR DESCRIPTION
**Pull request by zomars**
## Important note

For obvious reasons, the checks for this PR are going to fail since the required checks doesn't exist anymore. Here's a [link to the latest run on my fork](https://github.com/zomars/cal.com/actions/runs/4237316804) to prove it's working. After merge we should set the only required check to `pr.yml`.

## What does this PR do?

- Introduces a single job for PRs
- Conditionally runs embeds and App Store e2e tests if needed
- Adds proper usage of shared cache for turbo/yarn/node_modules and production builds
- Abstract common used tasks into [Composable Actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action)
- Transforms individual workflows into [Reusable Workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows)

image

## Type of change

- [x] Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

- [ ] View Checks on my [testing fork](https://github.com/zomars/cal.com/actions/runs/4237316804)

### CODE REVIEW

**Review by emrysal — APPROVED** [review]
Happy :)

**Comment by zomars** [comment]
Merging to make necessary adjustments on time.

File: .github/actions/cache-db/action.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+name: Cache databse between jobs
```
Comment (zomars) [inline .github/actions/cache-db/action.yml:1]:
```suggestion
name: Cache database between jobs
```

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -1,20 +1,11 @@
...
-concurrency:
-  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
-  cancel-in-progress: true
 jobs:
   e2e:
     timeout-minutes: 20
-    runs-on: buildjet-4vcpu-ubuntu-2204
-
+    # runs-on: buildjet-4vcpu-ubuntu-2204 # Uncomment before merging
+    runs-on: ubuntu-latest
```
Comment (zomars) [inline .github/workflows/e2e.yml]:
```suggestion
    runs-on: buildjet-4vcpu-ubuntu-2204
```

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -1,38 +1,15 @@
...
-      - "**.jsx?"
-concurrency:
-  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
-  cancel-in-progress: true
+  workflow_call:
 jobs:
   lint:
-    runs-on: buildjet-4vcpu-ubuntu-2204
+    # runs-on: buildjet-4vcpu-ubuntu-2204 # Uncomment before merging
+    runs-on: ubuntu-latest
```
Comment (zomars) [inline .github/workflows/lint.yml]:
```suggestion
    runs-on: buildjet-4vcpu-ubuntu-2204
```

**Pull request merged**